### PR TITLE
pythonPackages.easysnmp: init at 0.2.5; python3Packages.poster3: init at 0.8.1

### PR DESCRIPTION
**Pull request by WhittlesJr**
###### Motivation for this change
Add [easysnmp](https://github.com/kamakazikamikaze/easysnmp.git) and [poster3](https://pypi.org/project/poster3/). 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @lheckemann: whom I believe to be sphalerite on IRC, with whom I talked about this briefly and obliquely on IRC on 8/06/2019.

### CODE REVIEW

File: pkgs/development/python-modules/easysnmp/default.nix [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+, termcolor
+, ipython
+}:
+
+buildPythonPackage rec {
+  pname = "easysnmp";
+  version = "0.2.5";
+
+  # See https://github.com/kamakazikamikaze/easysnmp/issues/108
+  disabled = !(pythonOlder "3.7");
```
Comment (lheckemann) [inline pkgs/development/python-modules/easysnmp/default.nix]: `pythonAtLeast "3.7"`?

File: pkgs/development/python-modules/easysnmp/default.nix [hunk 2]
```diff
@@ -0,0 +1,72 @@
+{ stdenv
+, lib
+, pkgs
```
Comment (lheckemann) [inline pkgs/development/python-modules/easysnmp/default.nix]: remove this, and accept `net_snmp`, `openssl` etc as arguments instead

File: pkgs/development/python-modules/easysnmp/default.nix [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    sha256 = "1si9iyxqj6z22jzn6m93lwpinsqn20lix2py3jm3g3fmwawkd735";
+  };
+
+  checkInputs = [
+    pytest
+    pytestcov
+    pytest-flake8
+    pytest-sugar
+    termcolor
+    pkgs.net_snmp
```
Comment (lheckemann) [inline pkgs/development/python-modules/easysnmp/default.nix]: May not be necessary to repeat this from buildInputs, not sure.

File: pkgs/development/python-modules/easysnmp/default.nix [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  ];
+
+  buildInputs = [
+    pkgs.net_snmp
+    pkgs.openssl
+    pkgs.gcc
```
Comment (lheckemann) [inline pkgs/development/python-modules/easysnmp/default.nix]: I think gcc should be included as a result of this being built with stdenv anyway, not 100% sure though. If this builds without it, do remove it.

File: pkgs/development/python-modules/easysnmp/default.nix [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    ipython
```
Comment (lheckemann) [inline pkgs/development/python-modules/easysnmp/default.nix]: Is ipython really a dependency of easysnmp?

File: pkgs/development/python-modules/easysnmp/easysnmp.patch [hunk 6]
```diff
@@ -0,0 +1,57 @@
+diff --git a/setup.cfg b/setup.cfg
```
Comment (lheckemann) [inline pkgs/development/python-modules/easysnmp/easysnmp.patch]: Some explanation of why we need this patch would be good.

**Review by lheckemann — CHANGES_REQUESTED** [review]
You believe correctly ;)

Looks good overall, just a few changes.

**Comment by lheckemann** [comment]
@GrahamcOfBorg build python3Packages.easysnmp python3Packages.poster3

File: pkgs/development/python-modules/easysnmp/default.nix (hunk 4 above)
Comment (jonringer) [reply pkgs/development/python-modules/easysnmp/default.nix]: for darwin it's clang. However, more importantly, it should be inherited from stdenv.cc. This line should be unnecessary

File: pkgs/top-level/python-packages.nix [hunk 7]
```diff
@@ -1600,6 +1600,8 @@ in {
 
   dugong = callPackage ../development/python-modules/dugong {};
 
+  easysnmp = callPackage ../development/python-modules/easysnmp {};
```
Comment (jonringer) [inline pkgs/top-level/python-packages.nix]: you should avoid passing all pkgs directly, a good way to accidentally bloat your closure, just pass the packages you care about.
```suggestion
  easysnmp = callPackage ../development/python-modules/easysnmp { 
    openssl = pkgs.openssl; 
    net_snmp = pkgs.net_snmp;
  };
```

File: pkgs/development/python-modules/poster3/default.nix [hunk 8]
```diff
@@ -0,0 +1,40 @@
+{ lib
+, pkgs
```
Comment (jonringer) [inline pkgs/development/python-modules/poster3/default.nix]:
```suggestion
```

File: pkgs/development/python-modules/poster3/default.nix [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+, buildPythonPackage
+, fetchPypi
+, python
```
Comment (jonringer) [inline pkgs/development/python-modules/poster3/default.nix]:
```suggestion
```

File: pkgs/development/python-modules/poster3/default.nix [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+  checkInputs = [
+    paste
+    webob
+    pyopenssl
+  ];
+
+  doCheck = true;
+
+  meta = {
```
Comment (jonringer) [inline pkgs/development/python-modules/poster3/default.nix]: personal preference
```suggestion
  meta = with lib; {
```

File: pkgs/development/python-modules/poster3/default.nix [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    description = "Streaming HTTP uploads and multipart/form-data encoding";
+    homepage = https://atlee.ca/software/poster/;
+    license = lib.licenses.mit;
```
Comment (jonringer) [inline pkgs/development/python-modules/poster3/default.nix]:
```suggestion
    license = licenses.mit;
```

File: pkgs/development/python-modules/poster3/default.nix [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    maintainers = with lib.maintainers; [ WhittlesJr ];
```
Comment (jonringer) [inline pkgs/development/python-modules/poster3/default.nix]:
```suggestion
    maintainers = with maintainers; [ WhittlesJr ];
```

File: pkgs/development/python-modules/poster3/default.nix [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    sha256 = "1b27d7d63e3191e5d7238631fc828e4493590e94dcea034e386c079d853cce14";
+  };
+
+  checkInputs = [
+    paste
+    webob
+    pyopenssl
+  ];
+
+  doCheck = true;
```
Comment (jonringer) [inline pkgs/development/python-modules/poster3/default.nix]: this is defaulted to true in buildPythonPackage, you can remove
```suggestion
```

File: pkgs/development/python-modules/poster3/default.nix [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+, webob
+, pyopenssl
+}:
+
+buildPythonPackage rec {
+  pname = "poster3";
+  version = "0.8.1";
+  format = "wheel";
+
+  disabled = ! isPy3k;
```
Comment (jonringer) [inline pkgs/development/python-modules/poster3/default.nix]:
```suggestion
  disabled = !isPy3k;
```

File: pkgs/development/python-modules/easysnmp/default.nix [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+    python setup.py install --prefix=$out
+  '';
+
+  # I can't get tests to pass, even running by hand. This library needs maintenance.
+  doCheck = false;
+  #checkPhase = ''
+  #  python setup.py test
+  #'';
+
+  meta = {
```
Comment (jonringer) [inline pkgs/development/python-modules/easysnmp/default.nix]: personal preference
```suggestion
  meta = with lib; {
```

File: pkgs/development/python-modules/easysnmp/default.nix [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    description = "A blazingly fast and Pythonic SNMP library based on the official Net-SNMP bindings";
+    homepage = https://easysnmp.readthedocs.io/en/latest/;
+    license = lib.licenses.bsd3;
```
Comment (jonringer) [inline pkgs/development/python-modules/easysnmp/default.nix]:
```suggestion
    license = licenses.bsd3;
```

File: pkgs/development/python-modules/easysnmp/default.nix [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+    maintainers = with lib.maintainers; [ WhittlesJr ];
```
Comment (jonringer) [inline pkgs/development/python-modules/easysnmp/default.nix]:
```suggestion
    maintainers = with maintainers; [ WhittlesJr ];
```

File: pkgs/development/python-modules/easysnmp/default.nix [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    ipython
+  ];
+
+  patches = [./easysnmp.patch];
+
+  buildPhase = ''
+    python setup.py build --basedir=${pkgs.net_snmp}/bin
+  '';
+
+  installPhase= ''
```
Comment (jonringer) [inline pkgs/development/python-modules/easysnmp/default.nix]: This installPhase largely does what the default buildPythonPackage installPhase does, you should just remove this altogether.

File: pkgs/development/python-modules/easysnmp/default.nix [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+  ];
+
+  buildInputs = [
+    pkgs.net_snmp
+    pkgs.openssl
+    pkgs.gcc
+    ipython
+  ];
+
+  patches = [./easysnmp.patch];
```
Comment (jonringer) [inline pkgs/development/python-modules/easysnmp/default.nix]: since this patch is largely trying to fix tests, but the tests fail anyway, you should be able to remove

File: pkgs/development/python-modules/easysnmp/default.nix [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+    mkdir -p $dst
+    export PYTHONPATH="$dst:$PYTHONPATH"
+    python setup.py install --prefix=$out
+  '';
+
+  # I can't get tests to pass, even running by hand. This library needs maintenance.
+  doCheck = false;
+  #checkPhase = ''
+  #  python setup.py test
+  #'';
```
Comment (jonringer) [inline pkgs/development/python-modules/easysnmp/default.nix]: remove commented checkPhase

File: pkgs/development/python-modules/easysnmp/default.nix [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+  '';
+
+  installPhase= ''
+    dst=$out/${python.sitePackages}
+    mkdir -p $dst
+    export PYTHONPATH="$dst:$PYTHONPATH"
+    python setup.py install --prefix=$out
+  '';
+
+  # I can't get tests to pass, even running by hand. This library needs maintenance.
```
Comment (jonringer) [inline pkgs/development/python-modules/easysnmp/default.nix]: probably want to avoid the first person, and use the passive voice
```suggestion
  # Unable to get tests to pass, even running by hand. The pytest tests have become stale.
```

File: pkgs/top-level/python-packages.nix (hunk 7 above)
Comment (lheckemann) [reply pkgs/top-level/python-packages.nix]: No, pythonPackages's callPackage will fill stuff not found in pythonPackages from pkgs itself so that isn't necessary — it just needs to accept openssl and net_snmp.

File: pkgs/development/python-modules/easysnmp/default.nix (hunk 21 above)
Comment (lheckemann) [reply pkgs/development/python-modules/easysnmp/default.nix]: Why?

Comment (jonringer) [reply pkgs/development/python-modules/easysnmp/default.nix]: Seems odd reading it when you don't know who is saying it. idk. Don't feel strongly about this one.

File: pkgs/top-level/python-packages.nix (hunk 7 above)
Comment (jonringer) [reply pkgs/top-level/python-packages.nix]: if there's ever a package added to python-modules with those names, then it will break.

Comment (jonringer) [reply pkgs/top-level/python-packages.nix]: I'm fully aware that it will fall back to the pkgs namespace

File: pkgs/development/python-modules/easysnmp/default.nix (hunk 3 above)
Comment (risicle) [reply pkgs/development/python-modules/easysnmp/default.nix]: `checkInputs` should get all `buildInputs` too

File: pkgs/development/python-modules/poster3/default.nix (hunk 9 above)
Comment (WhittlesJr) [reply pkgs/development/python-modules/poster3/default.nix]: This is needed for accessing python.sitePackages. I see this done in a lot of other packages too. Is there a more preferred way?

File: pkgs/development/python-modules/easysnmp/default.nix (hunk 19 above)
Comment (WhittlesJr) [reply pkgs/development/python-modules/easysnmp/default.nix]: Shouldn't I leave it for posterity in case someone else wants to try to get tests to pass?

File: pkgs/development/python-modules/poster3/default.nix (hunk 9 above)
Comment (jonringer) [reply pkgs/development/python-modules/poster3/default.nix]: I have a lot of suggestions, if you use the default build/install phases, then it's not required.

File: pkgs/development/python-modules/easysnmp/default.nix (hunk 18 above)
Comment (WhittlesJr) [reply pkgs/development/python-modules/easysnmp/default.nix]: Doing so causes the following:
```
/nix/store/f74zj9535vgk9nlrskk1qgk3v6m7w6n4-stdenv-linux/setup: line 1324: pushd: dist: No such file or directory
```

File: pkgs/development/python-modules/easysnmp/default.nix (hunk 19 above)
Comment (jonringer) [reply pkgs/development/python-modules/easysnmp/default.nix]: maybe, but it looks like the current state was for an old version of pytest.... I would say it has bitrotted beyond intended use. 

The most proper thing would be to try and upstream the changes and just amend the next bump. But the repo seems like it's kind of inactive.

File: pkgs/development/python-modules/easysnmp/default.nix (hunk 18 above)
Comment (jonringer) [reply pkgs/development/python-modules/easysnmp/default.nix]: it's because you're overriding the buildPhase, buildPhase normall builds a wheel which would be `python setup.py build bdist_wheel`

this worked:
```

  buildPhase = ''
    python setup.py build bdist_wheel --basedir=${net_snmp}/bin
  '';

  # Unable to get tests to pass, even running by hand. The pytest tests have
  # become stale.
  doCheck = false;

```

Comment (WhittlesJr) [reply pkgs/development/python-modules/easysnmp/default.nix]: Ah yes, that does work!  Thank you.

**Comment by WhittlesJr** [comment]
I believe I've implemented all the feedback. Is the next step to squash and rebase onto master?

**Comment by jonringer** [comment]
yes, please rebase so that you have 3 PR's, 1 for maintainer entry, and 1 for each package. And it wouldn't be a bad time to rebase off master since this has been open a little while.

**Comment by WhittlesJr** [comment]
Oh, I was told to have three commits in one PR. I think that was @lheckemann's advice...

**Comment by WhittlesJr** [comment]
Oops hold on, I missed something

**Comment by jonringer** [comment]
sorry, my terms got mixed up, 3 commits in one PR

**Comment by WhittlesJr** [comment]
OK, all set, I think

**Review by jonringer — APPROVED** [review]
`nix-review` passes on NixOS
diff LGTM
leaf packages

a little concerned whether the easysnmp package needs the net_snmp bin at runtime, but I don't have a good way to verify if thats the case.

**Comment by jonringer** [comment]
@WhittlesJr if you address a comment, you can resolve the individual comment windows. :)

Should clean up chat quite a bit

**Comment by WhittlesJr** [comment]
@jonringer, I believe net_snmp is needed at runtime. Does it need to be propagatedBuildInputs instead of buildInputs? My usage of it has been working as written.

**Comment by jonringer** [comment]
I think in most practical cases, where they bring this package into an environment it should work.

The use case I was thinking of, is when a different package references this, then the path might not have the net_snmp binaries. Only way around this (that i know of) would be to patch the source so that it uniquely finds net_snmp. Example [patch](https://github.com/jonringer/nixpkgs/blob/e6223c022b94fad4b85e88319515d6f58910b5d7/pkgs/development/python-modules/pyproj/001.proj.patch) for the package [pyproj](https://github.com/jonringer/nixpkgs/blob/e6223c022b94fad4b85e88319515d6f58910b5d7/pkgs/development/python-modules/pyproj/default.nix#L19). This way, the package doesn't go to path or other mechanisms to find the binary, it will just return the full path exactly to what it's looking for, regardless of environment.

**Comment by WhittlesJr** [comment]
Hmm, after grepping around, I don't see it shelling out to any net-snmp binaries at runtime. I suspect we're probably good with what we have here.

**Review by jonringer — APPROVED** [review]
`nix-review` passes on NixOS
diff LGTM (Nice!)
leaf packages

```
https://github.com/NixOS/nixpkgs/pull/66645
2 package were build:
python27Packages.easysnmp python37Packages.poster3
```

**Comment by lheckemann** [comment]
One more nit: https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix#L23 was just introduced recently, could you add your github id to maintainers.nix as well? Then I'll merge :)

**Comment by WhittlesJr** [comment]
@lheckemann, how does that look? I also alphabetized my entry in maintainer-list.nix.

File: pkgs/top-level/python-packages.nix [hunk 22]
```diff
@@ -1608,6 +1608,11 @@ in {
 
   dugong = callPackage ../development/python-modules/dugong {};
 
+  easysnmp = callPackage ../development/python-modules/easysnmp {
+    openssl = pkgs.openssl;
```
Comment (lheckemann) [inline pkgs/top-level/python-packages.nix]: This shouldn't be necessary, pythonPackages will fill missing stuff in from pkgs.

**Comment by lheckemann** [comment]
Sorry, one more change! Other than that it does look good :sweat_smile:

File: pkgs/top-level/python-packages.nix (hunk 22 above)
Comment (jonringer) [reply pkgs/top-level/python-packages.nix]: i recommended this, in case there's a package added to python-modules of that name. I feel like it's worth-while future-proofing

Comment (WhittlesJr) [reply pkgs/top-level/python-packages.nix]: Makes sense to me

Comment (lheckemann) [reply pkgs/top-level/python-packages.nix]: Still :-1: on this. I'd be against adding something called `openssl` or `net_snmp` to python-packages anyway, and there are plenty of other packages using non-python deps without them being threaded through specifically. (also, if you really want to, you can rephrase this as `inherit (pkgs) openssl net_snmp`).

Comment (WhittlesJr) [reply pkgs/top-level/python-packages.nix]: I do see lots of examples of the `inherit (pkgs)` pattern. Shall we use it here?

Comment (jonringer) [reply pkgs/top-level/python-packages.nix]: I guess it's just because I'm not a big fan of bringing in a entire package sets. I wonder what @FRidh  has to say on pkgs vs per-package imports

Comment (WhittlesJr) [reply pkgs/top-level/python-packages.nix]: Rebased on master again. I'd like to do whatever we need to do to get this merged. Sounds like there's disagreement on which way to go, so we need an authoritative decision.

Comment (WhittlesJr) [reply pkgs/top-level/python-packages.nix]: Pinging @jonringer, @lheckemann, and @FRidh. I need to know which way to go with this change. Please advise.

Comment (jonringer) [reply pkgs/top-level/python-packages.nix]: I'm still in favor of the current state. I'm not a committer, so I can't merge :(

Comment (WhittlesJr) [reply pkgs/top-level/python-packages.nix]: @jonringer, do you know who else we might try pinging?

File: pkgs/development/python-modules/easysnmp/default.nix [hunk 23]
```diff
@@ -0,0 +1,57 @@
+{ stdenv
+, lib
+, buildPythonPackage
+, pythonAtLeast
+, fetchFromGitHub
+
```
Comment (jonringer) [inline pkgs/development/python-modules/easysnmp/default.nix]:
```suggestion
```

File: pkgs/development/python-modules/poster3/default.nix [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+, isPy3k
+, paste
+, webob
+, pyopenssl
+}:
+
+buildPythonPackage rec {
+  pname = "poster3";
+  version = "0.8.1";
+  format = "wheel";
```
Comment (jonringer) [inline pkgs/development/python-modules/poster3/default.nix]: add a comment about wheel format, it's the last option
```suggestion
  format = "wheel"; # only redistributable available
```

**Review by jonringer — COMMENTED** [review]
I'm not sure about adding easysnmp to nixpkgs, it looks like it's not maintained, and hasn't seen any attention for 13 months, https://github.com/kamakazikamikaze/easysnmp

**Comment by WhittlesJr** [comment]
That would be pretty discouraging... it's still actively used, if not maintained.

**Comment by lheckemann** [comment]
Sorry for the sluggishness!

**Comment by WhittlesJr** [comment]
Thank you!
